core/src/main/java/org/infinispan/persistence/file/SingleFileStore.java

```diff
@@ -918,51 +1063,131 @@ public boolean delete(Object key) {
       return entryFactory.create(keyBb, valueBb);
    }
 
+   /**
+    * @return The entries of a segment, or {@code null} if the segment is not owned
+    */
+   @GuardedBy("resizeLock")
+   private Map<K, FileEntry> getSegmentEntries(int segment) {
+      if (!segmented) {
+         return entries[0];
+      }
+
+      // Segmented
+      if (numSegments <= segment) {
+         throw new IndexOutOfBoundsException();
+      }
+      return entries[segment];
+   }
+
    @Override
-   public Flowable<K> publishKeys(Predicate<? super K> filter) {
-      return Flowable.fromIterable(() -> {
-         List<K> keys = new ArrayList<>(entries.size());
-         long now = ctx.getTimeService().wallClockTime();
-         synchronized (entries) {
-            for (Map.Entry<K, FileEntry> e : entries.entrySet()) {
+   public Publisher<K> publishKeys(IntSet segments, Predicate<? super K> filter) {
+      if (!segmented) {
+         return blockingManager.blockingPublisher(blockingPublishSegmentKeys(k -> keyMatches(segments, filter, k), 0));
+      }
+      return Flowable.fromIterable(segments)
+                     .concatMap(segment -> {
+                        Flowable<K> segmentPublisher = blockingPublishSegmentKeys(filter, segment);
```

Comment:
Sure, I can do the same optimization here, I just didn't think publishers are used often enough for it to make a difference.

The optimization only helps in a very particular scenario: passivation with all entries fitting in the data container, so there is no load from disk but there is a load for every cache miss, and a delete for every cache remove.

Reply:
My main concern is the first sentence, you are doing non blocking work in a blocking thread and possibly blocking work in a non blocking thread, opposite of what we want.

The optimization is more of a nice to have, I am more worried about the first thing really.

Reply:
I see what you mean now, I was using `blockingManager.blockingPublisher(segmentPublisher)` wrong on the next line because all the potentially blocking work was done in `blockingPublishSegmentKeys()`.